client/client-lib/src/api.rs

```diff
@@ -1,171 +1,575 @@
+use fedimint_api::task::{RwLock, RwLockWriteGuard};
+use fedimint_api::{dyn_newtype_define, NumPeers, PeerId, TransactionId};
 use fedimint_core::epoch::{SerdeEpochHistory, SignedEpochOutcome};
-use fedimint_core::modules::ln::contracts::incoming::IncomingContractOffer;
-use fedimint_core::modules::ln::contracts::ContractId;
-use fedimint_core::modules::ln::{ContractAccount, LightningGateway};
-use fedimint_core::modules::mint::db::ECashUserBackupSnapshot;
-use fedimint_core::modules::wallet::PegOutFees;
-use fedimint_core::outcome::legacy::{OutputOutcome, TryIntoOutcome};
 use fedimint_core::outcome::TransactionStatus;
-use fedimint_core::transaction::legacy::Transaction as LegacyTransaction;
 use fedimint_core::transaction::SerdeTransaction;
-use fedimint_core::CoreError;
 use futures::stream::FuturesUnordered;
-use futures::StreamExt;
+use futures::{Future, StreamExt};
 #[cfg(not(target_family = "wasm"))]
 use jsonrpsee_core::client::CertificateStore;
 use jsonrpsee_core::client::ClientT;
 use jsonrpsee_core::Error as JsonRpcError;
-use jsonrpsee_types::error::CallError as RpcCallError;
 #[cfg(target_family = "wasm")]
 use jsonrpsee_wasm_client::{Client as WsClient, WasmClientBuilder as WsClientBuilder};
 #[cfg(not(target_family = "wasm"))]
 use jsonrpsee_ws_client::{WsClient, WsClientBuilder};
-use serde::{Deserialize, Serialize};
+use serde::{de::DeserializeOwned, Deserialize, Serialize};
+use serde_json::Value;
 use thiserror::Error;
 use threshold_crypto::PublicKey;
-use tracing::{debug, error, instrument, trace, warn};
+use tokio::time::sleep;
+use tracing::{debug, error, instrument};
 use url::Url;
 
 use crate::module_decode_stubs;
-use crate::query::{
-    CurrentConsensus, EventuallyConsistent, QueryStep, QueryStrategy, Retry404, UnionResponses,
-    UnionResponsesSingle, ValidHistory,
-};
+use crate::query::{CurrentConsensus, QueryStep, QueryStrategy, UnionResponses, ValidHistory};
+use crate::LegacyTransaction;
 
 #[cfg_attr(target_family = "wasm", async_trait(? Send))]
 #[cfg_attr(not(target_family = "wasm"), async_trait)]
-pub trait IFederationApi: Debug + Send + Sync {
-    /// Fetch the outcome of an entire transaction
-    async fn fetch_tx_outcome(&self, tx: TransactionId) -> Result<TransactionStatus>;
+pub trait IFederationApi: Debug {
+    /// List of available federation members
+    fn members(&self) -> &BTreeSet<PeerId>;
+    /// Make request to a specific federation member by `peer_id`
+    async fn request_raw(&self, peer_id: PeerId, method: &str, params: &[Value]) -> Result<Value>;
+}
 
-    /// Submit a transaction to all federation members
-    async fn submit_transaction(&self, tx: LegacyTransaction) -> Result<TransactionId>;
+async fn with_erased_conversion<Params, Res, E, F>(params: &Params, f: F) -> result::Result<Res, E>
+where
+    Params: Serialize + Send + Sync,
+    Res: DeserializeOwned,
+    F: FnOnce(Vec<Value>) -> Pin<Box<dyn Future<Output = result::Result<Value, E>> + Send>> + Send,
```

Comment:
What would the conversion be used for?

Reply:
I can no longer view the code you anchored for some reason, so I have no idea which function was it.

Reply:
The reload code when you pushed the fixup might have messed it up, it was about the functions that took a closure and called it with a value. You changed them since then, so it's no longer relevant.